api/agent/state_trackers.go

```diff
@@ -109,11 +116,15 @@ func (c *containerState) UpdateState(ctx context.Context, newState ContainerStat
 
 	c.lock.Lock()
 
-	// except for 1) switching back to idle from busy (hot containers) or 2)
-	// to waiting from done, otherwise we can only move forward in states
+	// Only the following transitions are allowed:
+	// 1) from busy to idle/paused
+	// 2) from done to waiting
+	// 3) to/from idle from/to paused
+	// otherwise, we can only move forward in states
 	if c.state < newState ||
-		(c.state == ContainerStateBusy && newState == ContainerStateIdle) ||
-		(c.state == ContainerStateDone && newState == ContainerStateIdle) {
+		(c.state != newState && isIdleState(newState) && isIdleState(c.state)) ||
+		(c.state == ContainerStateBusy && isIdleState(newState)) ||
+		(c.state == ContainerStateDone && isIdleState(newState)) {
```

Comment:
i don't think this will allow done -> waiting (as noted, it should).

done < waiting -> false
done != waiting && isIdleState(done) && isIdleState(waiting) -> false, neither are idle states
done == busy && isIdleState(waiting) -> false, waiting is not an idle state
done == done && isIdleState(waiting) -> false, waiting is not an idle state

seemingly, making waiting an idle state would solve this, but didn't analyze the implications of that. my brain hurts now, bravo. at least, i think this needs some more kludging (or a tweak to the comments)

Reply:
done->waiting might be an old / deprecated requirement.. Checking... Thanks for catching this.